Userland/Utilities/sed.cpp

```diff
@@ -0,0 +1,559 @@
+/*
+ * Copyright (c) 2021, Rodrigo Tobar <rtobarc@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/CharacterTypes.h>
+#include <AK/Format.h>
+#include <AK/Optional.h>
+#include <AK/Tuple.h>
+#include <LibCore/ArgsParser.h>
+#include <LibCore/File.h>
+
+#define ENUMERATE_FUNCTIONS(F) \
+    F('a', 1)                  \
+    F('b', 2)                  \
+    F('c', 2)                  \
+    F('d', 2)                  \
+    F('D', 2)                  \
+    F('g', 2)                  \
+    F('G', 2)                  \
+    F('h', 2)                  \
+    F('H', 2)                  \
+    F('i', 1)                  \
+    F('l', 2)                  \
+    F('n', 2)                  \
+    F('N', 2)                  \
+    F('p', 2)                  \
+    F('P', 2)                  \
+    F('q', 1)                  \
+    F('r', 1)                  \
+    F('s', 2)                  \
+    F('t', 2)                  \
+    F('w', 2)                  \
+    F('x', 2)                  \
+    F('y', 2)                  \
+    F(':', 0)                  \
+    F('=', 1)                  \
+    F('#', 0)
+
+enum class AddressType {
+    Unset,
+    Line,
+    LastLine,
+    ContextAddress,
+};
+
+class Address {
+public:
+    Address() = default;
+
+    explicit Address(size_t line)
+        : m_line_number(line)
+        , m_address_type(AddressType::Line)
+    {
+    }
+
+    explicit Address(AddressType address_type)
+        : m_address_type(address_type)
+    {
+        VERIFY(address_type == AddressType::LastLine || address_type == AddressType::ContextAddress);
+    }
+
+    size_t line_number() const
+    {
+        VERIFY(m_address_type == AddressType::Line);
+        return m_line_number;
+    }
+
+    AddressType address_type() const { return m_address_type; }
+
+    bool matches(StringView /*pattern_space*/, size_t line_number, bool is_last_line) const
+    {
+        switch (m_address_type) {
+        case AddressType::Line:
+            return line_number == m_line_number;
+        case AddressType::LastLine:
+            return is_last_line;
+        default:
+            Core::File::standard_error()->write(String::formatted("Addressing type not implemented: {}", int(m_address_type)));
+            return false;
+        }
+    }
+
+private:
+    size_t m_line_number { 0 };
+    AddressType m_address_type { AddressType::Unset };
+};
+
+namespace AK {
+template<>
+class Formatter<Address> : public StandardFormatter {
+public:
+    void format(FormatBuilder& format_builder, Address address)
+    {
+        auto& builder = format_builder.builder();
+        switch (address.address_type()) {
+        case AddressType::Line:
+            builder.appendff("{}", address.line_number());
+            break;
+        case AddressType::LastLine:
+            builder.append('$');
+            break;
+        case AddressType::ContextAddress:
+            VERIFY_NOT_REACHED();
```

Comment:
Is this a `TODO()`?

Reply:
Mmm... it's both probably? This is the formatting routine for `Address`, but the execution can't reach this code at the moment because addresses of type `ContextAddress` are not yet parsed, so an `Address` with `ContextAddress` type is never constructed.

I'd say the TODO, if anywhere, would be more on the `parse_address` function, but if you think it's better to go with a `TODO()` here too I'm happy to change it.

Reply:
Either is fine, I just thought it might have been forgotten or something

Reply:
I'll leave the `VERIFY_NOT_REACHED` for the time being then.